### PR TITLE
Chapter 10: Updating, showing, and deleting users

### PR DESCRIPTION
**Pull request by baaohuii**

### CODE REVIEW

**Comment by baaohuii** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -5,11 +5,16 @@ def create
     user = User.find_by　email: params[:session][:email].downcase
     if user && user.authenticate　params[:session][:password]
       log_in user
+<<<<<<< HEAD
```
Comment (manhnt-0960) [inline app/controllers/sessions_controller.rb]: please fix conflict

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,4 +1,12 @@
 class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
```
Comment (manhnt-0960) [inline app/controllers/users_controller.rb]:
```suggestion
  before_action :logged_in_user, only: %i(index edit update destroy)
```

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,4 +1,12 @@
...
+  before_action :correct_user, only: [:edit, :update]
+  before_action :admin_user, only: :destroy
+
+  def index
+    @users = User.paginate page: params[:page]
```
Comment (manhnt-0960) [inline app/controllers/users_controller.rb]: add number of per page

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -23,9 +31,42 @@ def edit
     @user = User.find params[:id]
   end
 
+  def update
+    @user = User.find params[:id]
```
Comment (manhnt-0960) [inline app/controllers/users_controller.rb]: please use find_by method instead of find

File: app/views/users/_user.html.erb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<li>
+  <%= gravatar_for user, size: Settings.gravatar_size %>
+  <%= link_to user.name, user %>
+  <% if current_user.admin? && !current_user?(user) %>
+|   <%= link_to "delete", user, method: :delete,
+                          data: { confirm: "You sure?" } %>
+  <% end %>
+</li>
```
Comment (manhnt-0960) [inline app/views/users/_user.html.erb]: add newline at end of file

File: app/views/users/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (manhnt-0960) [inline app/views/users/index.html.erb]: remove empty line

**Comment by baaohuii** [comment]
ready

**Comment by baaohuii** [comment]
ready

File: Gemfile [hunk 7]
```diff
@@ -3,17 +3,20 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
...
+gem "faker"
 gem "jbuilder", "~> 2.7"
 gem "mysql2"
 gem "puma", "~> 4.1"
 gem "rails", "~> 6.0.3", ">= 6.0.3.2"
 gem "sass-rails", ">= 6"
 gem "turbolinks", "~> 5"
 gem "webpacker", "~> 4.0"
-gem "bcrypt"
+gem "will_paginate"
```
Comment (ngocvt-0484) [inline Gemfile]: please use gem kaminari

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -14,18 +22,52 @@ def create
...
     else
-      flash[:danger] = "Danger! Please try again."
+      flash[:danger] = "Sorry. Please try again."
       render :new
     end
   end
 
   def edit
-    @user = User.find params[:id]
+    @user = User.find_by params[:id]
```
Comment (ngocvt-0484) [inline app/controllers/users_controller.rb]: define a method load_user to find user. and use before_action to call it
EX:
before_action :load_user, only: [:edit, :update, :destroy]
def load_user
   @user = User.find_by params[:id]
end

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -14,18 +22,52 @@ def create
...
+  end
+
+  def update
+    @user = User.find_by params[:id]
```
Comment (ngocvt-0484) [inline app/controllers/users_controller.rb]: define a method load_user to find user. and use before_action to call it
EX:
before_action :load_user, only: [:edit, :update, :destroy]
def load_user
@user = User.find_by params[:id]
end

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -14,18 +22,52 @@ def create
...
+      flash[:success] = "Profile updated!"
+      redirect_to @user
+    else
+      flash.now[:danger] = "Update failed."
+      render :edit
+    end
+  end
+
+  def destroy
+    User.find_by params[:id].destroy
```
Comment (ngocvt-0484) [inline app/controllers/users_controller.rb]: define a method load_user to find user. and use before_action to call it
EX:
before_action :load_user, only: [:edit, :update, :destroy]
def load_user
@user = User.find_by params[:id]
end

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -14,18 +22,52 @@ def create
...
   end
 
   private
   
   def user_params
     params.require(:user).permit :name, :email, :password, :password_confirmation
   end
+
+  def logged_in_user
+    unless logged_in?
```
Comment (ngocvt-0484) [inline app/controllers/users_controller.rb]: return unless logged_in?

**Comment by baaohuii** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 12]
```diff
@@ -144,3 +144,13 @@ input.btn.btn-primary {
   width: auto;
   margin-left: 0;
 } 
+
+.users {
+  list-style: none;
+  margin: 0;
+  li {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  margin: 0;

  li {
```

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -1,6 +1,15 @@
...
+  before_action :correct_user, only: %i(edit update)
+  before_action :admin_user, only: :destroy
+
+  def index
+    @users = User.sort_by_name.page params[:page]
+  end
+
   def show
-  	@user = User.find_by id: params[:id]
+  	@user = User.find_by params[:id]
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
  	@user = User.find_by id: params[:id]
```

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -14,18 +23,60 @@ def create
       flash[:success] = "Welcome to the Sample App!"
       redirect_to @user
     else
-      flash[:danger] = "Danger! Please try again."
+      flash[:danger] = "Sorry. Please try again."
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
      flash.now[:danger] = "Sorry. Please try again."
```

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -14,18 +23,60 @@ def create
...
       render :new
     end
   end
 
   def edit
-    @user = User.find params[:id]
+    @user = User.find_by params[:id]
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    @user = User.find_by id: params[:id]
```

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -14,18 +23,60 @@ def create
...
+      flash[:success] = "Profile updated!"
+      redirect_to @user
+    else
+      flash.now[:danger] = "Update failed."
+      render :edit
+    end
+  end
+
+  def destroy
+    User.find_by params[:id].destroy
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    User.find_by(id: params[:id]).destroy
```

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -14,18 +23,60 @@ def create
...
-    params.require(:user).permit :name, :email, :password, :password_confirmation
+    params.require(:user).permit User::USERS_PARAMS
+  end
+
+  def logged_in_user
+    return unless logged_in?
+      store_location
+      flash[:danger] = "Please log in."
+      redirect_to login_url
+    end
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion

    store_location
    flash[:danger] = "Please log in."
    redirect_to login_url
```

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -14,18 +23,60 @@ def create
...
+  end
+  
+  def load_user
+    @user  = User.find_by params[:id]
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    @user  = User.find_by id: params[:id]
```

File: app/controllers/users_controller.rb [hunk 19]
```diff
@@ -14,18 +23,60 @@ def create
...
+    return if @user.present?
+
+    flash[:warning] = t "user.new.absent"
+    redirect_to root_path
+  end
+
+  def correct_user
+    @user = User.find_by params[:id]
+    redirect_to(root_url) unless current_user?(@user)
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    @user = User.find_by id: params[:id]
    return if current_user? @user
  
    redirect_to root_url
```

File: app/views/users/edit.html.erb [hunk 20]
```diff
@@ -0,0 +1,22 @@
+<% provide :title, "Edit user" %>
+<h1>Update your profile</h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
```
Comment (chungpv-1008) [inline app/views/users/edit.html.erb]:
```suggestion
    <%= form_for @user do |f| %>
```

File: app/views/users/edit.html.erb [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+      <%= f.email_field :email, class: "form-control" %>
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation, class: "form-control" %>
+      <%= f.submit "Save changes", class: "btn btn-primary" %>
+    <% end %>
+    <div class="gravatar_edit">
+      <%= gravatar_for @user %>
+      <a href= Settings.links.gravatar.emails target="_blank">change</a>
```
Comment (chungpv-1008) [inline app/views/users/edit.html.erb]: using link_to instead of tag 'a'

File: app/views/users/index.html.erb [hunk 22]
```diff
@@ -0,0 +1,10 @@
+<% provide :title, "All users" %>
+<h1>All users</h1>
+<%= paginate @users  %>
+<ul class="users">
+  <%= render @users %>
+  <% @users.each do |user| %>
+   <%= render user %>
```
Comment (chungpv-1008) [inline app/views/users/index.html.erb]:
```suggestion
  <%= render @users %>
```

**Comment by baaohuii** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
Try to refactor code such as Mrs Ngoc suggested

**Comment by baaohuii** [comment]
ready

File: app/controllers/users_controller.rb [hunk 23]
```diff
@@ -1,4 +1,13 @@
 class UsersController < ApplicationController
+  before_action :load_user, only: [:edit :update :destroy]
+  before_action :logged_in_user, only: [:show :new :create]
+  before_action :correct_user, only: [:edit :update]
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
  before_action :load_user, only: %i(edit update destroy)
  before_action :logged_in_user, only: %i(show new create)
  before_action :correct_user, only: %i(edit update)
```

File: app/controllers/users_controller.rb [hunk 24]
```diff
@@ -14,18 +23,57 @@ def create
...
     else
-      flash[:danger] = "Danger! Please try again."
+      flash.now[:danger] = "Sorry. Please try again."
       render :new
     end
   end
 
   def edit
-    @user = User.find params[:id]
+    @user = User.find_by id: params[:id]
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion

```

File: app/controllers/users_controller.rb [hunk 25]
```diff
@@ -14,18 +23,57 @@ def create
...
+  end
+
+  def update
+    @user = User.find_by params[:id]
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion

```

File: app/controllers/users_controller.rb [hunk 26]
```diff
@@ -14,18 +23,57 @@ def create
...
+      flash[:success] = "Profile updated!"
+      redirect_to @user
+    else
+      flash.now[:danger] = "Update failed."
+      render :edit
+    end
+  end
+
+  def destroy
+    User.find_by (id: params [:id]).destroy
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    User.find_by(id: params [:id]).destroy
```

File: app/controllers/users_controller.rb [hunk 27]
```diff
@@ -14,18 +23,57 @@ def create
...
   def user_params
-    params.require(:user).permit :name, :email, :password, :password_confirmation
+    params.require(:user).permit User::USERS_PARAMS
+  end
+
+  def logged_in_user
+    return unless logged_in?
+      store_location
+      flash[:danger] = "Please log in."
+      redirect_to login_url
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
    return unless logged_in?

    store_location
    flash[:danger] = "Please log in."
    redirect_to login_url
```

File: app/views/users/index.html.erb [hunk 28]
```diff
@@ -0,0 +1,10 @@
+<% provide :title, "All users" %>
+<h1>All users</h1>
+<%= paginate @users  %>
+<ul class="users">
+  <%= render @users %>
+  <% @users.each do |user| %>
+   <%= render @user %>
+  <% end %>
```
Comment (chungpv-1008) [inline app/views/users/index.html.erb]:
```suggestion
  <%= render @users %>
```

File: app/controllers/users_controller.rb (hunk 24 above)
Comment (baaohuii) [reply app/controllers/users_controller.rb]: Do you mean I need to delete the line32? or change to anything? I do not see what's the different between mine and suggested change

**Comment by baaohuii** [comment]
ready

File: app/controllers/users_controller.rb [hunk 29]
```diff
@@ -14,18 +23,55 @@ def create
...
-    params.require(:user).permit :name, :email, :password, :password_confirmation
+    params.require(:user).permit User::USERS_PARAMS
+  end
+
+  def logged_in_user
+    return unless logged_in?
+    
+      store_location
+      flash[:danger] = "Please log in."
+      redirect_to login_url
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
    return unless logged_in?
    
    store_location
    flash[:danger] = "Please log in."
    redirect_to login_url
```

File: app/controllers/users_controller.rb [hunk 30]
```diff
@@ -14,18 +23,55 @@ def create
...
+  end
+  
+  def load_user
+    @user  = User.find_by id: params[:id]
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb:64]:
```suggestion
    @user  = User.find_by id: params[:id]
    return if @user

    flash[:danger] = "Message"
    redirect_to root_url
```

File: app/controllers/users_controller.rb [hunk 31]
```diff
@@ -14,18 +23,55 @@ def create
...
+      flash[:success] = "Profile updated!"
+      redirect_to @user
+    else
+      flash.now[:danger] = "Update failed."
+      render :edit
+    end
+  end
+
+  def destroy
+    User.find_by(id: params [:id]).destroy
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
    @user.destroy
```

**Comment by baaohuii** [comment]
ready

**Review by chungpv-1008 — COMMENTED** [review]
merged

**Pull request merged**